otaclient/app/ota_client_stub.py

```diff
@@ -14,661 +14,720 @@
-            my_ecu_id=self.my_ecu_id,
+    def __init__(self, *, ecu_info: ECUInfo, _proxy_cfg=proxy_cfg):
+        self._executor = ThreadPoolExecutor(thread_name_prefix="otaclient_service_stub")
+        self._run_in_executor = partial(
+            asyncio.get_running_loop().run_in_executor, self._executor
         )
 
-        # for _get_subecu_status
-        self._status_pulling_lock = asyncio.Lock()
-        self._last_status_query = 0
-        self._cached_status = wrapper.StatusResponse()
+        self.ecu_info = ecu_info
+        self.listen_addr = ecu_info.ip_addr
+        self.listen_port = server_cfg.SERVER_PORT
+        self.my_ecu_id = ecu_info.ecu_id
 
-        # update session
-        self._update_session = _RequestHandlingSession(
-            enable_otaproxy=proxy_cfg.enable_local_ota_proxy,
+        self._otaclient_control_flags = OTAClientControlFlags()
+        self._otaclient_stub = OTAClientWrapper(
+            ecu_info=ecu_info,
             executor=self._executor,
+            control_flags=self._otaclient_control_flags,
+            proxy=_proxy_cfg.get_proxy_for_local_ota(),
         )
 
-    ###### API stub method #######
-    def host_addr(self):
-        return self._ecu_info.get_ecu_ip_addr()
-
-    async def update(self, request: wrapper.UpdateRequest) -> wrapper.UpdateResponse:
-        logger.debug(f"receive update request: {request}")
-        response = wrapper.UpdateResponse()
+        # ecu status tracking
+        self._ecu_status_storage = ECUStatusStorage(ecu_info)
+        self._ecu_status_tracker = _ECUTracker(
+            self._ecu_status_storage,
+            ecu_info=ecu_info,
+            otaclient_stub=self._otaclient_stub,
+        )
 
-        if self._update_session.is_running:
-            response.add_ecu(
-                wrapper.UpdateResponseEcu(
-                    ecu_id=self.my_ecu_id,
-                    result=wrapper.FailureType.RECOVERABLE,
-                )
-            )
-            logger.debug("ignore duplicated update request")
-            return response
+        # otaproxy lifecycle and dependency managing
+        if _proxy_cfg.enable_local_ota_proxy:
+            self._otaproxy_launcher = OTAProxyLauncher(executor=self._executor)
+            self._status_checking_shutdown_event = asyncio.Event()
+            asyncio.create_task(self._otaproxy_lifecycle_managing())
+            asyncio.create_task(self._otaclient_control_flags_managing())
         else:
-            # start this update session
-            _init_cache = (
-                self._ota_client.live_ota_status.get_ota_status()
-                == wrapper.StatusOta.SUCCESS
-            )
-            await self._update_session.start(request, init_cache=_init_cache)
+            # if otaproxy is not enabled, no dependency relationship will be formed,
+            # always allow local otaclient to reboot
+            self._otaclient_control_flags.set_can_reboot_flag()
```

Comment:
@obi-t4 Previously I think the dependency to otaproxy only caused by when we enable otaproxy, so we need to prevent otaclient reboots local ECU to keep otaproxy running. 
And if we don't enable otaproxy, the dependency will not be formed and otaclient is safe to reboot local ECU at anytime it wants.

But right now I am thinking of also let otaclient waits for all ECUs not in UPDATING ota_status, even we don't have dependency caused by requiring otaproxy, what do you think?

Reply:
@obi-t4 When otaproxy is not enabled(in proxy_info.yaml), otaclient and its child ECUs will not form dependency relationship during OTA update, so otaclient can reboot the local ECU once its own OTA is finished, without the need to wait for all child ECUs exit UPDATING ota_status. 
So we set the control flag to allow otaclient reboot at anytime.

Reply:
@Bodong-Yang 
In case of my ECU doesn't require otaproxy but some of the child ecu require otaproxy, what should we do?

Reply:
> @Bodong-Yang In case of my ECU doesn't require otaproxy but some of the child ecu require otaproxy, what should we do?

Actually I think this situation should not happen, this is configuration mistake and should be fixed by correcting `proxy_info.yaml` in main ECU.

If any child ECU requires otaproxy, main ECU should enable it.

Reply:
@obi-t4 

> In case of my ECU doesn't require otaproxy but some of the child ecu require otaproxy, what should we do?

For main ECU, not only itself, if any of its child ECU requires otaproxy, we should enable it. 
The situation that main ECU doesn't enable otaproxy(in `proxy_info.yaml`) but any of the child ECU requires it, should be considered as configuration mistake and should be fixed.

Reply:
( otaproxy enabled or not is determined by reading `proxy_info.yaml`'s configuration.